app/controllers/exercises_controller.rb

```diff
@@ -466,57 +467,58 @@ def reload
+            GROUP BY user_id, user_type;"
+    ApplicationRecord.connection.execute(query).each do |tuple|
+      user_statistics[tuple['user_type']][tuple['user_id'].to_i] = tuple
+    end
+    render locals: {
+      user_statistics: user_statistics,
+    }
+  end
+
+  def external_user_statistics
+    # Render statistics page for one specific external user
+
+    if policy(@exercise).detailed_statistics?
+      @submissions = Submission.where(user: @external_user,
+        exercise_id: @exercise.id).in_study_group_of(current_user).order('created_at')
+      @show_autosaves = params[:show_autosaves] == 'true'
+      @submissions = @submissions.where.not(cause: 'autosave') unless @show_autosaves
+      interventions = UserExerciseIntervention.where('user_id = ?  AND exercise_id = ?', @external_user.id,
+        @exercise.id)
+      @all_events = (@submissions + interventions).sort_by(&:created_at)
+      @deltas = @all_events.map.with_index do |item, index|
+        delta = item.created_at - @all_events[index - 1].created_at if index.positive?
+        delta.nil? || (delta > StatisticsHelper::WORKING_TIME_DELTA_IN_SECONDS) ? 0 : delta
+      end
+      @working_times_until = []
+      @all_events.each_with_index do |_, index|
+        @working_times_until.push((format_time_difference(@deltas[0..index].sum) if index.positive?))
       end
-      render 'exercises/external_users/statistics'
     else
-      # Show general statistic page for specific exercise
-      user_statistics = {'InternalUser' => {}, 'ExternalUser' => {}}
-      additional_filter = if policy(@exercise).detailed_statistics?
-                            ''
-                          elsif !policy(@exercise).detailed_statistics? && current_user.study_groups.count.positive?
-                            "AND study_group_id IN (#{current_user.study_groups.pluck(:id).join(', ')}) AND cause = 'submit'"
-                          else
-                            # e.g. internal user without any study groups, show no submissions
-                            'AND FALSE'
-                          end
-      query = "SELECT user_id, user_type, MAX(score) AS maximum_score, COUNT(id) AS runs
-              FROM submissions WHERE exercise_id = #{@exercise.id} #{additional_filter}
-              GROUP BY user_id, user_type;"
-      ApplicationRecord.connection.execute(query).each do |tuple|
-        user_statistics[tuple['user_type']][tuple['user_id'].to_i] = tuple
+      final_submissions = Submission.where(user: @external_user,
+        exercise_id: @exercise.id).in_study_group_of(current_user).final
+      @submissions = []
+      %i[before_deadline within_grace_period after_late_deadline].each do |filter|
+        relevant_submission = final_submissions.send(filter).latest
+        @submissions.push relevant_submission if relevant_submission.present?
       end
-      render locals: {
-        user_statistics: user_statistics,
-      }
+      @all_events = @submissions
```

Comment:
What is the reason for this assignment? Is the variable in this (prior) if branch just named differently? Can we align it?

Reply:
No, the variable is not _just_ named differently. Until now, we passed two instance variables to the view: `@submissions` consisting of all user submissions (and thus having files, ...) and `@all_events` consisting of the submissions plus all interventions that occurred for the learner. In the view, we then present a timeline with the submissions and interventions. Therefore, my proposal would be to eliminate the `@submissions` instance variable, and just pass `@all_events`. In the view, I can then separate those again with `@all_events.filter{|event| event.is_a? Submission}`. This is still not perfect (and I like the idea of integrating a presenter) but would reduce the number of variables passed to the view.

In the case we are currently looking at (a user not having the `policy(@exercise).detailed_statistics?`, we do not show the interventions at all (as a measurement to increase privacy). Therefore, only those are passed and the interventions don't need to be added. Compare that with the other line above:
https://github.com/openHPI/codeocean/blob/d28a3a0eaf6cffa030ad365a6716208af1e4e205/app/controllers/exercises_controller.rb#L501

Reply:
I made a proposal with my last commit 85f05712, please have look 😊

Reply:
Both options are not perfect (more variables vs. logic in the view). Typically, such logic like `submissions = @all_events.filter{|event| event.is_a? Submission}` should reside in a presenter. I don't mind too much.

Reply:
> Both options are not perfect (more variables vs. logic in the view)

I absolutely agree. Adding some presenters to the code base is also a nice suggestion, which I will consider for further refactoring. 

As discussed earlier today, I'll keep the new behaviour (with less variables) and continue with the other refactoring later. 

Thanks again for having a look :)